src/common/json.cc

```diff
@@ -0,0 +1,418 @@
+/*!
+ * Copyright 2018 by Contributors
+ * \file json.cc
+ * \brief JSON serialization of nested key-value store. Uses Tencent/RapidJSON
+ */
+#include "./json.h"
+#include <dmlc/logging.h>
+#include <rapidjson/istreamwrapper.h>
+#include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/reader.h>
+#include <rapidjson/writer.h>
+#include <istream>
+#include <stack>
+#include <limits>
+#include <unordered_map>
+
+namespace xgboost {
+namespace serializer {
+
+/* Use RapidJSON SAX API to parser JSON into nested key-value store. See
+ * http://rapidjson.org/md_doc_sax.html for more information. */
+class JSONInputHandler
+    : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, JSONInputHandler> {
+ public:
+  explicit JSONInputHandler(NestedKVStore* kvstore);
+  ~JSONInputHandler() = default;
+  bool Null();
+  bool Bool(bool b);
+  bool Int(int i);
+  bool Uint(unsigned u);
+  bool Int64(int64_t i);
+  bool Uint64(uint64_t u);
+  bool Double(double d);
+  bool String(const char* str, rapidjson::SizeType length, bool copy);
+  bool StartObject();
+  bool Key(const char* str, rapidjson::SizeType length, bool copy);
+  bool EndObject(rapidjson::SizeType memberCount);
+  bool StartArray();
+  bool EndArray(rapidjson::SizeType elementCount);
+
+  void CheckPostCondition();
+ private:
+  NestedKVStore* kvstore_;
+
+  enum class State : uint8_t {
+    kInit, kObject, kExpectValue, kExpectArrayItem
+  };
+  State parser_state_;
+  std::stack<NestedKVStore> object_context_;
+  std::string current_key_;
+  uint32_t parser_depth_;  // Depth of object currently being parsed (initially 0)
+};
+
+NestedKVStore LoadKVStoreFromJSON(std::istream* stream) {
+  NestedKVStore result;
+  JSONInputHandler handler(&result);
+  rapidjson::Reader reader;
+  rapidjson::IStreamWrapper isw(*stream);
+  CHECK(reader.Parse(isw, handler));
+  handler.CheckPostCondition();
+  return result;
+}
+
+template <typename OutputStreamType>
+void SaveKVStoreToJSON_(const NestedKVStore& kvstore,
```

Comment:
Maybe just let polymorphism take care of the naming. rapidjson::Writer is a different type than std::ostream.  Or maybe just combine it into one function, it feels reasonable to me.

Reply:
@trivialfis `SaveKVStoreToJSON_()` is a recursive function, so it needs to remain separate. But yes, it can be renamed to `SaveKVStoreToJSON()`.